### PR TITLE
feat: registration slug prefix manual set up

### PR DESCRIPTION
**Pull request by romanetar**
ref https://tipit.avaza.com/project/view#!tab=task-pane&groupby=MyTaskDueDate&view=vertical&task=3572465&stafilter=NotComplete&fileview=grid

Signed-off-by: romanetar <roman_ag@hotmail.com>

### CODE REVIEW

File: app/Http/Controllers/Apis/Protected/Summit/Factories/SummitValidationRulesFactory.php [hunk 1]
```diff
@@ -183,7 +183,7 @@ public static function buildForUpdate(array $payload = []): array
             'allow_update_attendee_extra_questions' => 'sometimes|boolean',
             'time_zone_label' => 'sometimes|string',
             'registration_allowed_refund_request_till_date' => 'nullable|date_format:U',
-            'registration_slug_prefix' => 'sometimes|string|max:50',
+            'registration_slug_prefix' => 'required|string|max:50',
```
Comment (smarcet) [inline app/Http/Controllers/Apis/Protected/Summit/Factories/SummitValidationRulesFactory.php:186]: on update is some times
not required

File: app/Models/Foundation/Summit/Summit.php [hunk 2]
```diff
@@ -2372,23 +2372,26 @@ public function getSlug($suffix = '-summit')
...
+    /**
+     * @param string $registration_slug_prefix
+     * @throws ValidationException
+     */
+    public function setRegistrationSlugPrefix(string $registration_slug_prefix)
+    {
+        // once that a ticket is sold, registration_slug_prefix can't be changed anymore
+        if ($this->getPaidTicketsCount() > 0)
+            throw new ValidationException("Cannot change the registration slug when there are paid tickets.");
+        $this->registration_slug_prefix = $registration_slug_prefix;
```
Comment (smarcet) [inline app/Models/Foundation/Summit/Summit.php:2392]: please do a trim and UPPER CASE HERE also REPLACE spaces BY _

File: app/Services/Model/Imp/SummitService.php [hunk 3]
```diff
@@ -1501,6 +1501,17 @@ public function addSummit(array $data)
...
             }
 
+            $registration_slug_prefix = $data['registration_slug_prefix'] ?? null;
+            if (!empty($registration_slug_prefix)) {
+                // check if exist another summit with that registration slug
+
+                $old_summit = $this->summit_repository->getByRegistrationSlugPrefix(trim($registration_slug_prefix));
+                if (!is_null($old_summit)) {
+                    throw new ValidationException(sprintf(
+                        "registration slug prefix %s already belongs to another summit", $registration_slug_prefix));
```
Comment (smarcet) [inline app/Services/Model/Imp/SummitService.php:1511]: fix the error message
should be
Registration slug prefix %s already belongs to another summit (%s).
where the second %s is the former summit id

File: app/Services/Model/Imp/SummitService.php [hunk 4]
```diff
@@ -1572,6 +1583,17 @@ public function updateSummit($summit_id, array $data)
...
```
Comment (smarcet) [inline app/Services/Model/Imp/SummitService.php:1593]: same as https://github.com/OpenStackweb/summit-api/pull/266/files#r1647583434

**Comment by smarcet** [comment]
@romanetar please review comments

File: app/Http/Controllers/Apis/Protected/Summit/Factories/SummitValidationRulesFactory.php (hunk 1 above)
Comment (romanetar) [reply app/Http/Controllers/Apis/Protected/Summit/Factories/SummitValidationRulesFactory.php:186]: done

File: app/Models/Foundation/Summit/Summit.php (hunk 2 above)
Comment (romanetar) [reply app/Models/Foundation/Summit/Summit.php:2392]: done

File: app/Repositories/Summit/DoctrineSummitRepository.php [hunk 5]
```diff
@@ -199,6 +199,26 @@ public function getBySlug(string $slug): ?Summit
...
+     * @return Summit|null
+     */
+    public function getByRegistrationSlugPrefix(string $registration_slug_prefix): ?Summit
+    {
+        try {
+            return $this->getEntityManager()->createQueryBuilder()
+                ->select("s")
+                ->from($this->getBaseEntity(), "s")
+                ->where('s.registration_slug_prefix = :registration_slug_prefix')
+                ->setParameter('registration_slug_prefix', strtolower($registration_slug_prefix))
```
[the remark this answers is not in the document]
Comment (romanetar) [reply app/Repositories/Summit/DoctrineSummitRepository.php:213]: done

File: app/Services/Model/Imp/SummitService.php [hunk 6]
```diff
@@ -1501,6 +1501,17 @@ public function addSummit(array $data)
                 }
             }
 
+            $registration_slug_prefix = $data['registration_slug_prefix'] ?? null;
+            if (!empty($registration_slug_prefix)) {
+                // check if exist another summit with that registration slug
+
+                $old_summit = $this->summit_repository->getByRegistrationSlugPrefix(trim($registration_slug_prefix));
```
[the remark this answers is not in the document]
Comment (romanetar) [reply app/Services/Model/Imp/SummitService.php:1508]: done

File: app/Services/Model/Imp/SummitService.php (hunk 3 above)
Comment (romanetar) [reply app/Services/Model/Imp/SummitService.php:1511]: done

File: app/Services/Model/Imp/SummitService.php [hunk 7]
```diff
@@ -1572,6 +1583,17 @@ public function updateSummit($summit_id, array $data)
...
```
[the remark this answers is not in the document]
Comment (romanetar) [reply app/Services/Model/Imp/SummitService.php:1590]: done

File: app/Services/Model/Imp/SummitService.php (hunk 4 above)
Comment (romanetar) [reply app/Services/Model/Imp/SummitService.php:1593]: done

File: app/Services/Model/Imp/SummitService.php [hunk 8]
```diff
@@ -1501,6 +1501,19 @@ public function addSummit(array $data)
                 }
             }
 
+            $registration_slug_prefix = $data['registration_slug_prefix'] ?? null;
```
Comment (smarcet) [inline app/Services/Model/Imp/SummitService.php:1504]: this is not correct
bc the SLUG get transformed when its set
so u need to externalize this

 $slug_prefix = strtoupper(trim(str_replace(" ","_", $registration_slug_prefix)));
on a static function and pass the user data by this function before doing the search

**Review by smarcet — CHANGES_REQUESTED** [review]
@romanetar please review comments

File: app/Services/Model/Imp/SummitService.php (hunk 8 above)
Comment (romanetar) [reply app/Services/Model/Imp/SummitService.php:1504]: done

**Review by smarcet — APPROVED** [review]
LGTM

**Pull request merged**